### PR TITLE
perf(moe): accelerate MXFP8 SiTU expert blocks on gfx950

### PR DESCRIPTION
**Pull request by qedawkins**
## Summary

Add an MXFP8-activation/MXFP4-weight Gluon expert pipeline for gfx950 EP8 through the existing MoE API. A single N16 weight bank serves both the grouped GEMMs and the A16 joint decoder for up to four rows. Use asynchronous stage-one loads, pipelined stage-two loads, sorted scales, and 32-row expert tiles through 1024 input rows. Small batches fuse expert mesh construction and counting; compact tiles fuse quantization with sorted-scale production. Larger inputs retain 128-row tiles.

Preserve checkpoint reloads and preallocated output storage, reject output views incompatible with packed BF16 atomics, and retain the existing TP and unsupported-shape paths.

## Testing Plan

Add reference-based coverage for quantization, weight/scale layouts, both GEMMs, SiTU, and weighted combination. Cover duplicate and remote routes, empty experts, ragged tiles, strided inputs/outputs, alignment rejection, and explicit versus automatic precision selection. Exercise changing-route graph replay, compiled-kernel reuse, the small-row joint decoder, and checkpoint reloads with TP slicing and stable parameter/storage identity.

## Benchmarks

MI355X; matched main versus this change.
Real Kimi-K3 weights and captured TP8/EP8 inputs: D=3584, I=3072, 112 local experts, top-k=16. Complete local routed-expert call, including sorting, quantization, both GEMMs, SiTU, and weighted combination; excludes router/top-k production, shared experts, latent projections, and collectives. Four-row controls are standalone experts, not the joint routed/shared call.

In-tree event timer, warm-cache ABBA, 100/1000 warmups/samples per block, no trimming. Values are averages of the two block medians; rank identifies the captured route distribution. Decode rows are 4*C with EAGLE3. These retained inputs are independent of the serving workload below.

```text
Phase        | Rows | Rank | Mode  | Main (us) | After (us) | Latency
-------------+------+------+-------+-----------+------------+--------
Prefill      | 1024 | 0    | eager | 842.43    | 485.80     | -42.3%
Prefill      | 1152 | 0    | eager | 809.53    | 668.49     | -17.4%
Prefill      | 8192 | 0    | eager | 2237.08   | 1269.50    | -43.3%
Prefill      | 8192 | 7    | eager | 2819.36   | 1538.43    | -45.4%
Decode       | 4    | 0    | graph | 57.68     | 78.10      | +35.4%
Decode       | 8    | 0    | graph | 97.56     | 60.42      | -38.1%
Decode       | 16   | 0    | graph | 125.02    | 71.60      | -42.7%
Decode       | 32   | 0    | graph | 170.30    | 94.42      | -44.6%
Decode       | 64   | 0    | graph | 237.40    | 134.25     | -43.4%
Decode       | 4    | 7    | graph | 34.12     | 40.79      | +19.5%
Decode       | 64   | 7    | graph | 189.02    | 104.56     | -44.7%
```

EvalScope streaming completions: 50K input / 500 output tokens, TP8/EP8, EAGLE3 (3 steps, 4 draft tokens, top-k 1), FP8 KV, no prefix cache/KVStore. Eager prefill with 8192-token chunks; decode graphs through batch 16. ABBA server restarts, two complete warmup ladders, 30 seconds idle, then a disjoint measured ladder; C requests per wave. Means of two measured waves per arm, no failures, exact input/output lengths. MLA is unchanged.

```text
C  | TTFT main/after (ms) | TPOT main/after (ms) | Output main/after (tok/s)
---+----------------------+----------------------+--------------------------
1  | 5233.38 / 4349.97    | 6.15 / 6.97          | 60.21 / 63.87
2  | 10316.94 / 8534.41   | 7.96 / 6.18          | 69.74 / 85.90
4  | 16717.26 / 15870.51  | 16.11 / 8.87         | 80.36 / 98.49
8  | 39756.18 / 32345.14  | 12.28 / 11.16        | 86.34 / 103.78
16 | 79834.48 / 65478.72  | 17.94 / 15.77        | 89.39 / 107.98
```

Output throughput includes prefill. Client TTFT/TPOT include admission and prefill interference, so their split is not an isolated decode-kernel timer. C1 TPOT regresses 13.4% despite identical speculative acceptance; the four-row N16 A16 controls also regress. Faster prefill does not remove this remaining small-row limitation.

### CODE REVIEW

File: tokenspeed-kernel-amd/README.md [hunk 1]
```diff
@@ -51,6 +51,168 @@ tokenspeed-kernel-amd/
 
 Public entry points currently remain architecture-specific. Consumers should import the implementation matching the target GPU, or use TokenSpeed-Kernel to select a compatible implementation through its registry.
 
+## MXFP8 SiTU Experts
```
Comment (antiagainst) [inline tokenspeed-kernel-amd/README.md]: Let's keep the top-level `README.md` for high level package information. If you'd like to log some details on op kernel support, follow the structure and put in to `python/tokenspeed-kernel-amd/ops/README.md`..

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/sched_barrier_c472597c64c30b9604155f08eb72d512b0f0e8873cd3d5ab6445e3d6b9f4079b.ll [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+; IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+; FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+; AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+; LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+; OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+; SOFTWARE.
+
+declare void @llvm.amdgcn.sched.barrier(i32 immarg)
+
+define i32 @__tokenspeed_sched_v1_barrier0() alwaysinline {
```
Comment (antiagainst) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/sched_barrier_c472597c64c30b9604155f08eb72d512b0f0e8873cd3d5ab6445e3d6b9f4079b.ll]: I'm wondering why the long file name and why the curious `v1` inside?

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/mxfp8_quantize.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        + (base_k[:, None] + column[None, :]).to(gl.int64) * INNER,
+        mask=row[:, None] < ROWS,
+        other=0.0,
+    ).to(gl.float32)
+    # Sequential fmax ignores NaNs; the floor also covers an all-NaN group.
+    magnitudes = gl.where(values == values, gl.abs(values), 0.0)
+    amax = gl.maximum(gl.max(magnitudes, 1), 1.0e-10)
+    bits = (amax * (1.0 / 448.0)).to(gl.uint32, bitcast=True)
+    exponent = (bits >> 23) & 255
+    exponent += ((exponent < 255) & ((bits & 0x7FFFFF) != 0)).to(gl.uint32)
```
Comment (antiagainst) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/mxfp8_quantize.py:66]: We have a lot of manual math here. Would [`cdna4.scaled_downcast`](https://triton-lang.org/main/gluon/api/generated/triton.experimental.gluon.language.amd.cdna4.scaled_downcast.html#triton.experimental.gluon.language.amd.cdna4.scaled_downcast) work here somehow? See https://github.com/lightseekorg/tokenspeed/pull/1442 for example. (I didn't check the algorithm carefully so just curious.)

File: tokenspeed-kernel/test/amd/ops/moe/test_mxfp8_contracts.py
Comment (antiagainst) [inline tokenspeed-kernel/test/amd/ops/moe/test_mxfp8_contracts.py:1]: We should move the sched barrier and this test into some common place. But can happen in the next step if you prefere.

File: test/runtime/layers/test_latent_moe_mxfp8_gfx950.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+import torch
+from torch import nn
+
+
+@pytest.fixture(scope="module")
+def prepared_latent_experts():
+    from tokenspeed_kernel.ops.moe import latent_moe_decode_pipeline_available
+    from tokenspeed_kernel.platform import current_platform
+
+    if not current_platform().is_cdna4:
```
Comment (antiagainst) [inline test/runtime/layers/test_latent_moe_mxfp8_gfx950.py]: Should we do module level skip if not meeting platform expectations?

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/_schedule.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    digest = hashlib.sha256(path.read_bytes()).hexdigest()
+    if path.name != f"sched_barrier_{digest}.ll":
+        raise RuntimeError(
+            "MXFP8 scheduler library must use its content-addressed filename"
+        )
+    return str(path)
+
+
+def _mxfp8_compile_options(*, enable_asan: bool) -> dict:
+    # The AMD ASAN linker supplies its own libraries and ignores extern_libs.
```
Comment (antiagainst) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/_schedule.py]: We can ignore asan--not sure we will really use it.

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/_schedule.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+_SCHED_SYMBOL = "__tokenspeed_sched_v1_barrier0"
+_SCHED_LIBRARY_PATH = str(
+    Path(__file__).with_name(
+        "sched_barrier_c472597c64c30b9604155f08eb72d512b0f0e8873cd3d5ab6445e3d6b9f4079b.ll"
+    )
+)
+
+
+@lru_cache(maxsize=1)
+def _checked_scheduler_library() -> str:
```
Comment (antiagainst) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/_schedule.py]: Is this really necessary? What is it guarding against?

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/expert_mesh.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+from tokenspeed_kernel_amd._triton import gl, gluon, triton
+
+
+@gluon.jit
+def _add(a, b):
+    return a + b
+
+
+@gluon.jit
+def _popcount(bits):
```
Comment (antiagainst) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/expert_mesh.py]: Is this just `libdevice.popc`?

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/mxfp8_gemm.py [hunk 8]
```diff
@@ -0,0 +1,743 @@
...
+    scales_a, scales_b = _load_scales(sa, sb, K, 0, BM)
+    _phase_boundary(SCHED_BARRIER)
+    if K > 256:
+        _copy_a(a1, x, offsets, 1, BUFFER_SAFE)
+    b = ()
+    for kh in gl.static_range(2):
+        for ni in gl.static_range(2):
+            b += (_load_b(w, K, 0, kh, ni),)
+    cdna4_async_copy.wait_group(0)
+    gl.barrier()
```
Comment (antiagainst) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/mxfp8_gemm.py]: Do we need these `gl.barrier` (here and everywhere elese)? If depdency detected the compiler will automatically inject them?

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/mxfp8_gemm.py [hunk 9]
```diff
@@ -0,0 +1,743 @@
...
+            (b_word >> 24).to(gl.uint8),
+            "e2m1",
+            acc[mi][1],
+        )
+        updated += ((c0, c1),)
+    return updated
+
+
+@gluon.jit
+def _rcp(x):
```
Comment (antiagainst) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/mxfp8_gemm.py]: `libdevice.fast_dividef`?

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/prefill_mxfp8.py
Comment (antiagainst) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/prefill_mxfp8.py:1]: This file name is not very consistent with others existing now, like `prefill_stageN.py`. Probably want to uniform them?

**Review by antiagainst — CHANGES_REQUESTED** [review]
Neat!

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/sched_barrier_c472597c64c30b9604155f08eb72d512b0f0e8873cd3d5ab6445e3d6b9f4079b.ll (hunk 2 above)
Comment (qedawkins) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/sched_barrier_c472597c64c30b9604155f08eb72d512b0f0e8873cd3d5ab6445e3d6b9f4079b.ll]: On the nvidia side they hash the extern libs file contents: https://github.com/triton-lang/triton/blob/b63e34c521dfc55a5f9e419cfeaa64df8263891f/third_party/nvidia/backend/compiler.py#L164-L168

but for some reason we don't on the HIP side: https://github.com/triton-lang/triton/blob/b63e34c521dfc55a5f9e419cfeaa64df8263891f/third_party/amd/backend/compiler.py#L270-L278

Probably should just change it so we do the same and then I can simplify the filename here.

Comment (qedawkins) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/sched_barrier_c472597c64c30b9604155f08eb72d512b0f0e8873cd3d5ab6445e3d6b9f4079b.ll]: Simplified filename and will just risk the cache issue for now.

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/mxfp8_quantize.py (hunk 3 above)
Comment (qedawkins) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/mxfp8_quantize.py:66]: scaled_downcast works, done

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/_schedule.py (hunk 5 above)
Comment (qedawkins) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/_schedule.py]: Dropped

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/mxfp8_gemm.py (hunk 8 above)
Comment (qedawkins) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/mxfp8_gemm.py]: Dropping all of them appears to have worked out, done

File: test/runtime/layers/test_latent_moe_mxfp8_gfx950.py [hunk 10]
```diff
@@ -26,16 +26,16 @@
 
 import pytest
 import torch
+from tokenspeed_kernel.platform import current_platform
 from torch import nn
 
+if not current_platform().is_cdna4:
```
Comment (antiagainst) [inline test/runtime/layers/test_latent_moe_mxfp8_gfx950.py]: We have `is_cdna4` in `tokenspeed-kernel/test/utils.py`

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/sched_barrier_c472597c64c30b9604155f08eb72d512b0f0e8873cd3d5ab6445e3d6b9f4079b.ll (hunk 2 above)
Comment (antiagainst) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/moe/mxfp4/sched_barrier_c472597c64c30b9604155f08eb72d512b0f0e8873cd3d5ab6445e3d6b9f4079b.ll]: I see. Yeah we should fix. Can you send a pull request against triton?

**Review by antiagainst — DISMISSED** [review]
Looks great, thanks!

File: test/runtime/layers/test_latent_moe_mxfp8_gfx950.py (hunk 10 above)
Comment (qedawkins) [reply test/runtime/layers/test_latent_moe_mxfp8_gfx950.py]: Hmmm, this is under `test/runtime/layers` which makes the import awkward.

Comment (antiagainst) [reply test/runtime/layers/test_latent_moe_mxfp8_gfx950.py]: I see. Missed it's under runtime path. Ignore my initial comments then
